Revise test expectations

piotr-iohk opened the pull request:
# Issue Number

#1238

# Overview

- 0a19dd650666ba3ca53b3e76da7d92a117fa0802
  Remove expectations:
  - expectFieldEqual  
  - expectFieldNotEqual
  - expectListItemFieldEqual
  - expectFieldBetween
  - expectListItemFieldBetween

- 074544e3b09d27bb51dd1917e626146988951e3c
  Add expectations:
  - expectCliFieldSatisfy
  - expectCliListItemFieldSatisfy

  and remove:
  - expectCliFieldBetween
  - expectCliFieldEqual
  - expectCliFieldNotEqual
  - expectCliListItemFieldBetween
  - expectCliListItemFieldEqual

- 7cf0c88575b8367c78ac0f9709dba9625544fb57
  Use eventually_ to remove:
  - expectEventually
  - expectEventually'
  - expectEventuallyL

- 2b2c51dc3a4775d88033c6e97d2c4c73f90bf332
  Rename expect.. methods to make them more concise

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Commit:
Remove expectations:

- expectFieldEqual  
- expectFieldNotEqual
- expectListItemFieldEqual
- expectFieldBetween
- expectListItemFieldBetween

Commit:
Add expectations:

- expectCliFieldSatisfy
- expectCliListItemFieldSatisfy

and remove:

- expectCliFieldBetween
- expectCliFieldEqual
- expectCliFieldNotEqual
- expectCliListItemFieldBetween
- expectCliListItemFieldEqual

Commit:
Use eventually_ to remove:

- expectEventually
- expectEventually'
- expectEventuallyL

Anviking commented on lib/core-integration/src/Test/Integration/Framework/DSL.hs:
```diff
@@ -530,62 +372,30 @@ expectValidJSON _ str =
-expectCliListItemFieldBetween
-    :: (HasCallStack, MonadIO m, MonadFail m, Show a, Eq a, Ord a)
+expectCliListItemFieldSatisfy
+    :: (MonadIO m, MonadFail m, Show a)
```
Let's keep `HasCallStack` such that the failures point to the source location where `expectCliFieldSatisfy` was called, instead of to this definition!

You kept it in the other ones:
```
 src/Test/Integration/Scenario/API/Transactions.hs:195:5:
  1) API Specifications TRANS_CREATE_01 - Single Output Transaction
       uncaught exception: IOException of type UserError
       user error (predicate failed for: 24)

  To rerun use: --match "/API Specifications/TRANS_CREATE_01 - Single Output Transaction/"

Randomized with seed 1239180884
```

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs:
```diff
@@ -207,37 +204,38 @@ spec = do
-            , expectFieldEqual (#direction . #getApiT) Outgoing
-            , expectFieldEqual (#status . #getApiT) Pending
+            , expectFieldSatisfy (#amount . #getQuantity) (>= feeMin + amt)
+            , expectFieldSatisfy (#amount . #getQuantity) (<= feeMax + amt)
```
The direction here is very nice. But unfortunately the failure-descriptions are now unhelpful:
```
Failures:

  src/Test/Integration/Scenario/API/Transactions.hs:195:5:
  1) API Specifications TRANS_CREATE_01 - Single Output Transaction
       uncaught exception: IOException of type UserError
       user error (predicate failed for: 24)

  To rerun use: --match "/API Specifications/TRANS_CREATE_01 - Single Output Transaction/"
```

I think we need to capture the same `between` we had before while keeping it as composable as you have now, by making the predicate return failure reasons. (Not just `a -> Bool`)

At the call-site, maybe it could look like this:

```haskell
expectFieldSatisfy 
    (#amount . #getQuantity) 
    (between (feeMin + amt, feeMax + amt))
```
or some variant as you prefer

_actually, the above would read as a sentence if expectFieldSatisfy was named expectField or expect_ 🤔

piotr-iohk replied:
That is a fair point, for the current `expectFieldBetween` the failulre message is:

```
  src/Test/Integration/Scenario/API/Transactions.hs:198:5: 
  1) API Specifications TRANS_CREATE_01x - Single Output Transaction
       uncaught exception: IOException of type UserError
       user error (expected 24 <= 23)

```

So we are missing the expected condition now.

I think the idea of #1238 was, for now, to replace exisitng expactations with `expect...Satisfy` to make DSL lighter.
On top of those changes we may start introducing improvements. (in subsequent PRs)

Good point also on makeing the methods shorter. I have change that in -> 2b2c51dc3a4775d88033c6e97d2c4c73f90bf332
I also like the idea of `between` and having:

```
expectField #amount (between (feeMin, feeMax))
``` 
Which, indeed, looks neat. 

Again, I would postpone it to the subsequent PRs. Thoughts?

Anviking replied:
> Again, I would postpone it to the subsequent PRs. Thoughts?

My concern would be that failing tests will be more difficult to debug before we fix that.

I also suspect it would be easier to go straight from `expectFieldBetween` to `expectField #foo (between (a, b))`, than converting to two `expectFieldSatisfy` checks (`>=`, and `<=`)

Anviking replied:
I think we would only need:
- `equal` - we can re-use `Test.Hspec.Expectations.shouldBe`
- `notEqual` - we can re-use `Test.Hspec.Expectations.shouldNotBe`
- `between :: (a, a) -> a -> IO ()` the only one we need to implement ourselves

Commit:
Rename expect.. methods to make them more concise

Anviking committed:
Replace (a -> Bool) predicates with (a -> Expectation)

For better failure reasons.

Note inconsistent naming (TODO: Decide on something)
- greaterThan
- (.>=)

Also note that there remains a few uses of (a -> Bool) via
shouldSatisfy. They require some un-straight-forward parenteses, which
could maybe be done nicer.

Commit:
git rebase master
